packages/cli/src/modules/workflow-index/workflow-index.service.ts

```diff
@@ -0,0 +1,165 @@
+import { Logger } from '@n8n/backend-common';
+import { WorkflowDependency, WorkflowDependencyRepository, WorkflowRepository } from '@n8n/db';
+import { Service } from '@n8n/di';
+import { IWorkflowBase } from 'n8n-workflow';
+
+import { EventService } from '@/events/event.service';
+
+@Service()
+export class WorkflowIndexService {
+	constructor(
+		private readonly eventService: EventService,
+		private readonly dependencyRepository: WorkflowDependencyRepository,
+		private readonly workflowRepository: WorkflowRepository,
+		private readonly logger: Logger,
+	) {}
+
+	init() {
+		this.eventService.on('server-started', async () => {
+			this.logger.info('Building workflow dependency index...');
+			// eslint-disable-next-line @typescript-eslint/return-await
+			return this.buildIndex();
+		});
+		this.eventService.on('workflow-created', async ({ workflow }) => {
+			await this.updateIndexFor(workflow);
+		});
+		this.eventService.on('workflow-saved', async ({ workflow }) => {
+			await this.updateIndexFor(workflow);
+		});
+		this.eventService.on('workflow-deleted', async ({ workflowId }) => {
+			// Updating with an empty list effectively removes the existing dependencies.
+			await this.dependencyRepository.updateDependenciesForWorkflow(workflowId, []);
+		});
+		this.eventService.on('workflow-archived', async ({ workflowId }) => {
+			// Updating with an empty list effectively removes the existing dependencies.
+			await this.dependencyRepository.updateDependenciesForWorkflow(workflowId, []);
+		});
+		this.eventService.on('workflow-unarchived', async ({ workflowId }) => {
+			await this.workflowRepository.findOneBy({ id: workflowId }).then(async (workflow) => {
+				if (workflow) {
+					await this.updateIndexFor(workflow);
+				}
+			});
+		});
+	}
+
+	private async buildIndex() {
+		// Get all the workflows.
+		const workflows = await this.workflowRepository.find();
+		// Build the index for each workflow.
+		for (const workflow of workflows) {
+			await this.updateIndexFor(workflow);
+		}
+	}
+
+	/**
+	 * Update the dependency index for a given workflow.
+	 *
+	 * NOTE: this should generally be handled via events, rather than called directly.
+	 * The exception is during workflow imports where it's simpler to call directly.
+	 *
+	 * @param workflow
+	 */
+	async updateIndexFor(workflow: IWorkflowBase) {
+		// TODO: input validation.
+		// Generate the dependency updates for the given workflow.
+		const dependencyUpdates: WorkflowDependency[] = [];
+
+		this.addNodeTypeDependencies(workflow, dependencyUpdates);
+		this.addCredentialDependencies(workflow, dependencyUpdates);
+		this.addWorkflowCallDependencies(workflow, dependencyUpdates);
+		this.addWebhookPathDependencies(workflow, dependencyUpdates);
```

Comment:
This is pretty clean and well separated. However if there are hundreds of of nodes (some users unfortunately do that), we iterate over them again for every dependency type. This might feel like premature optimization, but at least for the initial indexing it can add up. Since each dependency operates on a single node, maybe we can do loop over the nodes once and call each dependency with the node. This works as long as we don't have other type of dependencies